README.md
# Jitter Frame Format

A simple, easy to parse and robust serial frame format.
Intended for reliable data transmission in embedded applications, for example communication between a microcontroller and host device via UART or similar low-level protocols.

## Scope

The protocol defines a data frame-format, but is not a full network stack:

- No higher-level concepts such as data streams or delivery guarantees
- No requirements of specific electrical interface protocols (e.g. it can work over serial, RS485, or something else)

## Features

### Clear distinction between packet header and payload data

The protocol must be able to handle any payload data without making assumptions about 'unlikely' or 'special' data patterns such as:

- pseudorandom data (i.e. noisy sensor data, encrypted data)
- long sequences of ones or zeroes
- untrusted user-input (worst-case: crafted to confuse/exploit protocol parsers)

 This is implemented by base-64 encoding the payload data, while the packet headers contain some bit patterns that cannot be represented in base-64. This guarantees that payload data can never be mistaken for packet header data (and vice versa).

### Robust recovery from invalid data / FIFO overflow

The protocol is designed to guarantee recovery from:

- temporarily not keeping up with incoming data
- receiving partial packets (timeout)
- hot-plugging while another device is already sending data

These situations may occasionally happen due to logic or timing behavior, even on platforms where the electrical layer is perfectly noise free.

The incoming data stream can be syncronized to the start of a new frame by scanning for a particular bit pattern that only occurs once in a valid packet.

### Detection of data corrupt

The protocol is targeted for short-range communication where random bit-flips / noise is relatively rare: A 16-bit CRC checksum is used to detect corruption of payload data.

If your application expect strong electrical interference (e.g. long-range, wireless transmission), a protocol with more more elaborate checksums or error-correcting encoding may be more appropriate.

## Packet format

![Packet data fields](uart_protocol_bits.drawio.png)

The protocol defines a packet containing a `Header` and `Data` section. Each packet may carry up to `46269` bytes of data and is tagged with a `ID` value in the range `0x0000-0xF0FF`. Note that all fields use little-endian byte ordering.

### Header

The header contains four fields:

- `0xF1`: marks *start-of-frame*
- `ID` (u16, little endian): identifies the type of content in this frame. Can be any value `0` - `0xF0FF`.
- `Length`(u16, little endian): length of the data section (`0` - `0xF0FF`). Note that this length includes the protocol overhead. The maximum size of actual payload data is `(0xF0FF * 6)/8 - 2 = 46269` bytes
- `0xFF`: marks *end-of-header*

The *start-of-frame* / *end-of-header* values make it easy to guarantee correct parsing when recovering from data corruption. The values are chosen to meet the following criteria:

- outside the range of valid base-64
- must not occur in the high byte of the `ID` or `Length` field (for maximum robustness when syncronizing). `0xFF` is a convenient value as it reduces the range of those fields the least.
- different from each other: `0xF1` still has minimum impact on the range of `ID` / `Length` and at the same time encodes the protocol version (V1)

### Data

The data section contains two fields:

- `Payload`: any sequence of up to 46269 bytes (zero length is also valid)
- `CRC16`(u16, little endian): CRC16 checksum over the payload. The CRC16-USB algorithm is used.

The complete data section (including CRC16) is base64-encoded with the standard base64 charset (`A-Za-z0-9+/`).

While base64 encoding comes with a small overhead, it comes with a big advantage: arbitrary binary data can be encapsulated without risk of confusing it for a packet header (the base64 character set does not cover `0xF1` / `0xFF`).

## Syncronizing to incoming data at arbitrary offset

This section attempts to visualize how the protocol guarantees (re-)syncronization to incoming data frames (which may happen when hot-plugging devices, recovering from corruption, FIFO overflow etc).

![synchronizing incoming data](uart_protocol.drawio.png)

The figure compares an expected packet with incoming data at all possible byte offsets. Highlighted in red are the fields where the parser can detect a mismatch. Note that there is only one offset where the data is valid: this holds for all valid values of `ID` and `Length`. In all other cases the mismatch is detected by:
- 0xF1 != 0xFF
- 0xF1 cannot occur in second byte of `ID`, `Length` or base64 data
- 0xFF cannot occur in second byte of `ID`, `Length` or base64 data


CHANGELOG.md

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

-

## [0.1.1] - 2022-06-14

### Fixed

- Off by one: A packet with size 31 (chunk size + 1) whould result in a panic


src/receiver.rs
use crate::{
    error::Error,
    frame::{self, FrameHeader, START_OF_FRAME},
};

pub struct Receiver<RX> {
    bytes_skipped: u32,
    rx: RX,
}

pub trait ReceiveQueue {
    fn bytes_available(&self) -> usize;
    fn peek_at(&self, offset: usize) -> Option<u8>;
    fn flush(&mut self, n_bytes: usize);
}

impl<RX> Receiver<RX>
where
    RX: ReceiveQueue,
{
    pub fn new(rx: RX) -> Self {
        Self {
            rx,
            bytes_skipped: 0,
        }
    }

    /// Returns total amount of incoming bytes that were discarded
    ///
    /// The receiver only discards bytes that cannot form a valid frame.
    /// This could happen when recovering after a connection loss / corrupt packet / hotplug situation.
    /// This is a statistic similar to 'packet loss' in networking: substantial amount of bytes are skipped
    /// may indicate a bad link quality
    pub fn bytes_skipped(&self) -> u32 {
        self.bytes_skipped
    }

    fn peek_bytes(
        &self,
        offset: usize,
        n: usize,
        result: &mut [u8],
    ) -> nb::Result<(), crate::error::Error> {
        for i in 0..n {
            match self.rx.peek_at(offset + i) {
                Some(byte) => {
                    result[i] = byte;
                }
                None => {
                    return Err(nb::Error::Other(Error::QueueUnderflow));
                }
            }
        }
        Ok(())
    }

    fn skip_byte(&mut self) {
        self.rx.flush(1);
        self.bytes_skipped += 1;
    }

    fn rx_header(&mut self) -> nb::Result<FrameHeader, crate::error::Error> {
        // Skip bytes untill START_OF_RAME is detected
        loop {
            match self.rx.peek_at(0) {
                None => return Err(nb::Error::WouldBlock),
                Some(START_OF_FRAME) => {
                    break;
                }
                _ => {
                    self.skip_byte();
                }
            }
        }

        // Wait untill enough data is available to form a packet header
        if self.rx.bytes_available() < 6 {
            return Err(nb::Error::WouldBlock);
        }

        // Build a packet header
        let mut header_bytes = [0_u8; 6];
        self.peek_bytes(0, 6, &mut header_bytes[0..])?;
        match FrameHeader::try_from(header_bytes) {
            Ok(header) => Ok(header),
            Err(e) => {
                // Header invalid: skip a byte and try again next time...
                self.skip_byte();
                return Err(nb::Error::Other(e));
            }
        }
    }

    pub fn receive<const N: usize>(&mut self) -> nb::Result<frame::Frame<N>, crate::error::Error> {
        let header = self.rx_header()?;

        // Incoming packet would be too long to receive correctly: discard it
        if header.payload_len() > N {
            self.skip_byte();
            return Err(nb::Error::Other(Error::TooManyBytes));
        }

        // Wait untill enough data is available to form a complete packet
        let total_len = header.total_packet_len();
        if self.rx.bytes_available() < total_len {
            return Err(nb::Error::WouldBlock);
        }

        let mut data = [0_u8; N];
        let data_len = header.data_len();
        self.peek_bytes(6, data_len, &mut data[0..])?;
        match frame::Frame::try_from((header, &data[0..data_len])) {
            Ok(frame) => {
                self.rx.flush(total_len);
                Ok(frame)
            }
            Err(e) => {
                // Frame invalid: skip a byte and try again next time...
                self.skip_byte();
                Err(nb::Error::Other(e))
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::frame::{Frame, END_OF_HEADER, START_OF_FRAME};

    use super::{ReceiveQueue, Receiver};

    struct DummyReceiver<'a> {
        data: &'a [u8],
        rx_count: &'a mut usize,
    }
    impl ReceiveQueue for DummyReceiver<'_> {
        fn bytes_available(&self) -> usize {
            self.data.len() - *self.rx_count
        }

        fn peek_at(&self, offset: usize) -> Option<u8> {
            let read_offset = *self.rx_count;
            if offset < self.data.len() {
                Some(self.data[read_offset + offset])
            } else {
                panic!("DummyReceiver: peek past end of data!");
            }
        }

        fn flush(&mut self, n_bytes: usize) {
            *self.rx_count += n_bytes;
        }
    }

    #[test]
    fn receive_works() {
        let mut data = [
            START_OF_FRAME,
            0x37,
            0x13,
            0x07,
            0x00,
            END_OF_HEADER,
            0x41,
            0x41,
            0x45,
            0x43,
            0x44,
            0x6D,
            0x34,
        ];
        let mut rx_count: usize = 0;
        let rx = DummyReceiver {
            data: &mut data,
            rx_count: &mut rx_count,
        };
        let mut receiver = Receiver::new(rx);
        let frame: Frame<128> = receiver.receive().expect("Receive failed!");

        assert_eq!(0x1337, frame.id());
        assert_eq!(3, frame.bytes().len());
        assert_eq!(0, frame.bytes()[0]);
        assert_eq!(1, frame.bytes()[1]);
        assert_eq!(2, frame.bytes()[2]);
        assert_eq!(0, receiver.bytes_skipped());
    }

    #[test]
    fn receive_empty_works() {
        let mut data = [START_OF_FRAME, 0x37, 0x13, 0x00, 0x00, END_OF_HEADER];
        let mut rx_count: usize = 0;
        let rx = DummyReceiver {
            data: &mut data,
            rx_count: &mut rx_count,
        };
        let mut receiver = Receiver::new(rx);
        let frame: Frame<128> = receiver.receive().expect("Receive failed!");

        assert_eq!(0x1337, frame.id());
        assert_eq!(0, frame.bytes().len());
        assert_eq!(0, receiver.bytes_skipped());
    }

    #[test]
    fn receive_skip1_works() {
        let data = [
            0x34,
            START_OF_FRAME,
            0x37,
            0x13,
            0x07,
            0x00,
            END_OF_HEADER,
            0x41,
            0x41,
            0x45,
            0x43,
            0x44,
            0x6D,
            0x34,
        ];
        let mut rx_count: usize = 0;
        let rx = DummyReceiver {
            data: &data,
            rx_count: &mut rx_count,
        };
        let mut receiver = Receiver::new(rx);
        let frame: Frame<128> = receiver.receive().expect("Receive failed!");

        assert_eq!(0x1337, frame.id());
        assert_eq!(3, frame.bytes().len());
        assert_eq!(0, frame.bytes()[0]);
        assert_eq!(1, frame.bytes()[1]);
        assert_eq!(2, frame.bytes()[2]);
        assert_eq!(1, receiver.bytes_skipped());
    }

    #[test]
    fn receive_skip_any_works() {
        let data = [
            0x37,
            0x13,
            0x07,
            0x00,
            END_OF_HEADER,
            0x41,
            0x41,
            0x45,
            0x43,
            0x44,
            0x6D,
            0x34,
            START_OF_FRAME,
            0x37,
            0x13,
            0x07,
            0x00,
            END_OF_HEADER,
            0x41,
            0x41,
            0x45,
            0x43,
            0x44,
            0x6D,
            0x34,
        ];

        for offset in 0..12 {
            let mut test_data = data;
            test_data[offset] = START_OF_FRAME;
            let test_data = &test_data[offset..];
            let mut rx_count: usize = 0;
            let rx = DummyReceiver {
                data: test_data,
                rx_count: &mut rx_count,
            };
            assert_eq!(START_OF_FRAME, test_data[0]);
            let mut receiver = Receiver::new(rx);
            let _e = receiver
                .receive::<128>()
                .expect_err("Invalid data should be skipped");
            let frame = receiver.receive::<128>().expect("Valid");

            assert_eq!(0x1337, frame.id());
            assert_eq!(3, frame.bytes().len());
            assert_eq!(0, frame.bytes()[0]);
            assert_eq!(1, frame.bytes()[1]);
            assert_eq!(2, frame.bytes()[2]);
        }
    }
}


src/frame.rs
use crate::error::Error;
use crc::{Crc, CRC_16_USB};

#[derive(Debug)]
pub struct Frame<const N: usize> {
    header: FrameHeader,
    data: [u8; N],
}

#[derive(Debug)]
pub(crate) struct FrameHeader {
    id: u16,
    length: u16, // NB: length of base64-data
}

pub const START_OF_FRAME: u8 = 0xF1;
pub const END_OF_HEADER: u8 = 0xFF;
pub const ID_MAX: u16 = 0xF0FF;
pub const LENGTH_MAX: u16 = 0xF0FF;

fn div_round_up(a: usize, b: usize) -> usize {
    if a == 0 {
        return 1;
    }

    (a - 1) / b + 1
}

impl FrameHeader {
    fn calculate_length_field(payload_length: usize) -> Result<u16, Error> {
        // Calculate size used when encoding the given data as a Frame:
        // 6-byte header, 2-byte CRC, base64 overhead

        // No payload: there won't be any CRC or base64 overhead
        if payload_length == 0 {
            return Ok(0);
        }

        // prevent int overflow in calculation
        if payload_length >= ((usize::MAX / 8) - 2) {
            return Err(Error::InvalidLength);
        }
        let packet_length = div_round_up((payload_length + 2) * 8, 6);
        if packet_length > LENGTH_MAX as usize {
            return Err(Error::InvalidLength);
        }
        Ok(packet_length as u16)
    }

    /// Create a FrameHeader, calculating length field based on payload_length
    pub fn new(id: u16, payload_length: usize) -> Result<Self, Error> {
        let length = Self::calculate_length_field(payload_length)?;

        Self::from_raw(id, length)
    }

    /// Create a FrameHeader from raw field values
    fn from_raw(id: u16, length: u16) -> Result<Self, Error> {
        if id > ID_MAX {
            return Err(Error::InvalidID);
        }
        if length > LENGTH_MAX {
            return Err(Error::InvalidLength);
        }

        Ok(Self { id, length })
    }

    pub fn data_len(&self) -> usize {
        self.length as usize
    }

    pub fn total_packet_len(&self) -> usize {
        6 + self.data_len()
    }

    pub fn payload_len(&self) -> usize {
        // base64 to binary: 6 bits per character
        let binary_len = self.data_len() * 6 / 8;

        if binary_len >= 2 {
            // excluding 2-byte CRC
            binary_len - 2
        } else {
            // No payload data
            0
        }
    }

    pub fn as_bytes(self: Self) -> [u8; 6] {
        let id_bytes: [u8; 2] = self.id.to_le_bytes();
        let len_bytes: [u8; 2] = self.length.to_le_bytes();
        [
            START_OF_FRAME,
            id_bytes[0],
            id_bytes[1],
            len_bytes[0],
            len_bytes[1],
            END_OF_HEADER,
        ]
    }
}

impl TryFrom<[u8; 6]> for FrameHeader {
    type Error = Error;

    fn try_from(slice: [u8; 6]) -> Result<Self, Self::Error> {
        Self::try_from(&slice[0..6])
    }
}

impl TryFrom<&[u8]> for FrameHeader {
    type Error = Error;

    fn try_from(slice: &[u8]) -> Result<Self, Self::Error> {
        if slice.len() != 6 {
            return Err(Error::TooFewBytes);
        }
        if slice.len() > 6 {
            return Err(Error::TooManyBytes);
        }

        // Parse start-of-frame marker
        if slice[0] != START_OF_FRAME {
            return Err(Error::InvalidHeader);
        }

        // Parse ID
        let id_bytes: [u8; 2] = slice[1..3].try_into().map_err(|_| Error::TooFewBytes)?;
        let id = u16::from_le_bytes(id_bytes);

        // Parse length
        let len_bytes: [u8; 2] = slice[3..5].try_into().map_err(|_| Error::TooFewBytes)?;
        let length = u16::from_le_bytes(len_bytes);

        // Parse end-of-header marker
        if slice[5] != END_OF_HEADER {
            return Err(Error::InvalidHeader);
        }

        Self::from_raw(id, length)
    }
}

impl<const N: usize> Frame<N> {
    pub fn new(id: u16, payload: &[u8]) -> Result<Self, Error> {
        let header = FrameHeader::new(id, payload.len())?;

        Ok({
            // pre-initialize
            let mut s = Self {
                header,
                data: [0; N],
            };

            // copy data
            for (i, byte) in payload.iter().enumerate() {
                s.data[i] = *byte;
            }

            s
        })
    }

    pub fn id(&self) -> u16 {
        self.header.id
    }

    pub fn bytes(&self) -> &[u8] {
        &self.data[0..self.header.payload_len()]
    }
}

// try_from header + slice
impl<const N: usize> TryFrom<(FrameHeader, &[u8])> for Frame<N> {
    type Error = Error;

    fn try_from(header_and_bytes: (FrameHeader, &[u8])) -> Result<Self, Self::Error> {
        let (header, b64_data) = header_and_bytes;

        if N < header.payload_len() {
            return Err(Error::TooManyBytes);
        }

        let b64_len = header.data_len();
        if b64_data.len() < b64_len {
            return Err(Error::TooFewBytes);
        }
        if b64_data.len() > b64_len {
            return Err(Error::TooManyBytes);
        }

        let mut frame = Self {
            header,
            data: [0; N],
        };

        // No data to decode: frame is done
        if b64_len == 0 {
            return Ok(frame);
        }

        // Last few bytes may not fit in the output buffer as the encoded data contain 2 extra bytes of CRC checksum.
        // In base64 this is not guaranteed to be at a byte boundary, so we have to decode the last few bytes of data carefully!
        let split_offset = if b64_len < 8 {
            0
        } else {
            (b64_len - 4) & !3 // boundary at multiple of 4: 4 characters decode into exactly 3 bytes
        };

        // Decode bulk of the data directly into frame
        let base64_cfg = base64::Config::new(base64::CharacterSet::Standard, false);
        let bulk_decoded_size =
            base64::decode_config_slice(&b64_data[0..split_offset], base64_cfg, &mut frame.data)
                .map_err(|_| Error::InvalidBase64)?;

        // Decode last few bytes including CRC checksum
        let mut last_bytes: [u8; 8] = [0; 8];
        let remaining_len =
            base64::decode_config_slice(&b64_data[split_offset..], base64_cfg, &mut last_bytes)
                .map_err(|_| Error::InvalidBase64)?;

        if remaining_len < 2 {
            return Err(Error::TooFewBytes);
        }

        // Copy remaining data to frame
        let remaining_data_len = remaining_len - 2;
        let total_data_len = bulk_decoded_size + remaining_data_len;
        assert!(total_data_len == frame.header.payload_len());
        for (i, byte) in last_bytes[0..remaining_data_len].iter().enumerate() {
            frame.data[bulk_decoded_size + i] = *byte;
        }

        // Parse CRC
        let crc_bytes = &last_bytes[remaining_data_len..remaining_len];
        let crc_bytes: [u8; 2] = crc_bytes.try_into().map_err(|_| Error::TooFewBytes)?;
        let parsed_crc = u16::from_le_bytes(crc_bytes);

        // Verify CRC
        // CRC16 checksum is calculated over all binary payload data
        let crc = Crc::<u16>::new(&CRC_16_USB);
        let mut checksum = crc.digest();
        let len = frame.header.payload_len();
        checksum.update(&frame.data[..len]);
        let checksum = checksum.finalize();

        if parsed_crc != checksum {
            return Err(Error::InvalidCRC);
        }

        Ok(frame)
    }
}

// try_from slice
impl<const N: usize> TryFrom<&[u8]> for Frame<N> {
    type Error = Error;

    fn try_from(slice: &[u8]) -> Result<Self, Self::Error> {
        let header: FrameHeader = slice[0..6].try_into()?;
        let b64_data = &slice[6..];

        Self::try_from((header, b64_data))
    }
}

// try_from ref to array
impl<const N: usize, const L: usize> TryFrom<&[u8; L]> for Frame<N> {
    type Error = Error;

    fn try_from(value: &[u8; L]) -> Result<Self, Self::Error> {
        let slice: &[u8] = value;
        Self::try_from(slice)
    }
}

#[cfg(test)]
mod tests {
    use super::{Frame, END_OF_HEADER, START_OF_FRAME};
    use crate::error::Error;

    fn valid_frame_bytes() -> [u8; 13] {
        [
            // Frame header
            START_OF_FRAME, // Start-of-frame marker
            0x37,           // packet ID 0x1337 as little-endian (low byte)
            0x13,           // packet ID 0x1337 as little-endian (high byte)
            0x07,           // Packet length 7 (4-byte data + 3-byte CRC) (low byte)
            0x00,           // Packet length 7 (4-byte data + 3-byte CRC) (high byte)
            END_OF_HEADER,  // End-of-header marker
            // base64-encoded [00, 01, 02] should be "AAEC" = [0x41, 0x41, 0x45, 0x43]
            0x41,
            0x41,
            0x45,
            0x43,
            // CRC16-USB over [00, 01, 02] should be 0x6E0E = [0x0E, 0x6E] (little-endian) = "Dm4"
            0x44,
            0x6D,
            0x34,
        ]
    }

    #[test]
    fn valid_new() {
        // Should be a valid frame containing 3 bytes
        let frame: Frame<3> = Frame::new(0x1337, &[0, 1, 2]).expect("Valid frame");
        assert_eq!(0x1337, frame.id());
        assert_eq!(3, frame.bytes().len());
        assert_eq!(0, frame.bytes()[0]);
        assert_eq!(1, frame.bytes()[1]);
        assert_eq!(2, frame.bytes()[2]);
    }

    #[test]
    fn valid_from_bytes() {
        let frame = valid_frame_bytes();

        // Should be a valid frame containing 3 bytes
        let _frame: Frame<3> = Frame::try_from(&frame).expect("Valid frame");
    }

    #[test]
    fn invalid_start_of_frame_from_bytes() {
        let mut frame = valid_frame_bytes();
        frame[0] = 0xF2; // invalid start-of-frame

        let err = Frame::<128>::try_from(&frame).expect_err("Should not be a valid frame header");
        assert_eq!(Error::InvalidHeader, err);
    }
    #[test]
    fn invalid_end_of_header_from_bytes() {
        let mut frame = valid_frame_bytes();
        frame[5] = START_OF_FRAME; // invalid end-of-header

        let err = Frame::<128>::try_from(&frame).expect_err("Should not be a valid frame header");
        assert_eq!(Error::InvalidHeader, err);
    }

    #[test]
    fn invalid_id_from_bytes() {
        let mut frame = valid_frame_bytes();
        frame[2] = START_OF_FRAME; // invalid ID: MSB cannot go >= 0xF0

        let err = Frame::<128>::try_from(&frame).expect_err("Should not be a valid frame header");
        assert_eq!(Error::InvalidID, err);
    }

    #[test]
    fn invalid_length_from_bytes() {
        let mut frame = valid_frame_bytes();
        frame[4] = START_OF_FRAME; // invalid Length: MSB cannot go >= 0xF0

        let err = Frame::<128>::try_from(&frame).expect_err("Should not be a valid frame header");
        assert_eq!(Error::InvalidLength, err);
    }

    #[test]
    fn invalid_length2_from_bytes() {
        let mut frame = valid_frame_bytes();
        frame[3] = 6; // wrong length: actual data is 7 bytes

        let err = Frame::<128>::try_from(&frame).expect_err("Should not be a valid frame header");
        assert_eq!(Error::TooManyBytes, err);
    }

    #[test]
    fn frame_too_small_from_bytes() {
        let frame = valid_frame_bytes();

        // Frame defined impossibly small
        let err = Frame::<1>::try_from(&frame).expect_err("Should not be a valid frame header");
        assert_eq!(Error::TooManyBytes, err);
    }

    #[test]
    fn frame_too_small2_from_bytes() {
        let frame = valid_frame_bytes();

        // Frame defined one byte too small
        let err = Frame::<2>::try_from(&frame).expect_err("Should not be a valid frame header");
        assert_eq!(Error::TooManyBytes, err);
    }

    #[test]
    fn invalid_crc_from_bytes() {
        let mut frame = valid_frame_bytes();
        frame[6] = 0x42; // corrupt first byte

        let err = Frame::<128>::try_from(&frame).expect_err("CRC should mismatch!");
        assert_eq!(Error::InvalidCRC, err);
    }
    #[test]
    fn invalid_crc2_from_bytes() {
        let mut frame = valid_frame_bytes();
        frame[9] = 0x42; // corrupt last byte

        let err = Frame::<128>::try_from(&frame).expect_err("CRC should mismatch!");
        assert_eq!(Error::InvalidCRC, err);
    }
    #[test]
    fn invalid_crc3_from_bytes() {
        let mut frame = valid_frame_bytes();
        frame[11] = 0x42; // corrupt CRC byte

        let err = Frame::<128>::try_from(&frame).expect_err("CRC should mismatch!");
        assert_eq!(Error::InvalidCRC, err);
    }

    #[test]
    fn invalid_base64_from_bytes() {
        let mut frame = valid_frame_bytes();
        frame[11] = 0x80; // invalid base64 character

        let err = Frame::<128>::try_from(&frame).expect_err("CRC should mismatch!");
        assert_eq!(Error::InvalidBase64, err);
    }
}
